### PR TITLE
Custom gpu ops partitioning

### PR DESCRIPTION
**Pull request by keshavb96**
PR to update tutorial to demonstrate the use of custom_partitioning instead of xmap when writing a custom primitive in JAX

### CODE REVIEW

File: docs/Custom_Operation_for_GPUs.md [hunk 1]
```diff
@@ -34,7 +34,7 @@ You need to follow these steps in Python:
 * Define its abstract evaluation.
 * Define its lowering to MLIR.
 * [Optional] Define the gradient.
-* [Optional] Use [xmap](https://jax.readthedocs.io/en/latest/notebooks/xmap_tutorial.html) (or one of the experimental [custom_partitioning](https://jax.readthedocs.io/en/latest/jax.experimental.custom_partitioning.html) or [shard_map](https://jax.readthedocs.io/en/latest/jep/14273-shard-map.html) functions) for fast multi-GPU.
+* [Optional] Use [xmap](https://jax.readthedocs.io/en/latest/notebooks/xmap_tutorial.html), [custom_partitioning](https://jax.readthedocs.io/en/latest/jax.experimental.custom_partitioning.html) or [shard_map](https://jax.readthedocs.io/en/latest/jep/14273-shard-map.html) functions for fast multi-GPU.
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:37]: Move xmap to the end and say it is deprecated?

File: docs/Custom_Operation_for_GPUs.md [hunk 2]
```diff
@@ -558,543 +558,313 @@ The values have been computed correctly for forward operation, however, the gene
 
 As XLA does not have enough knowledge about the custom functions to shard input tensors, it decides to replicate them to produce correct values before making the custom call.
 
-To avoid this overhead, we need to use the xmap manual sharding with the following configuration updates
+To avoid this duplication, we can:
+- [custom_partitioning](https://jax.readthedocs.io/en/latest/jax.experimental.custom_partitioning.html): to make it behave like all native JAX operations (but more complicated)
+- Use manual sharding
+  - [xmap](https://jax.readthedocs.io/en/latest/notebooks/xmap_tutorial.html): deprecated and bugged in some cases when combined with grad
+  - [shard_map](https://jax.readthedocs.io/en/latest/jep/14273-shard-map.html): experimental but should work
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:565]: Remove experimental?

File: docs/Custom_Operation_for_GPUs.md [hunk 3]
```diff
@@ -558,543 +558,313 @@ The values have been computed correctly for forward operation, however, the gene
...
-        out_axes=("x", None, None, None),
-        axis_resources={"x": "x"},
-    )
-    reshaped_out = xmapped(reshaped, weight)
-    return reshaped_out.reshape(x.shape)
-```
+### Shard the forward function with custom_partitioning
 
-Now we need to run `xmap_rms_norm`, not `rms_norm` through `pjit`.
+When using custom_partitioning, it isn't needed to define the JAX config experimental_xmap_spmd_lowering and experimental_xmap_spmd_lowering_manual as was the case with the now deprecated xmap.
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:571]: As we don't show xmap anymore, remove.

File: docs/Custom_Operation_for_GPUs.md [hunk 4]
```diff
@@ -558,543 +558,313 @@ The values have been computed correctly for forward operation, however, the gene
...
-```
 ```python
-True
+def register_primitive(cls):
+    """
+    register jax primitive
+
+    The order of calls.
+
+    Inner, only the basic to wrap the TE XLA(not JAX) custom_call itself.
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:582]: Remove TE in this and lines around.

File: docs/Custom_Operation_for_GPUs.md [hunk 5]
```diff
@@ -558,543 +558,313 @@ The values have been computed correctly for forward operation, however, the gene
...
+    - Impl to TE XLA custom_call in C.
+    - abstract to know the static shapes
+    - lower to StableHLO TE XLA custom_call.
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:585]: No abreviation (Impl). Make it consistent with a capital at the start and a period at the end.

File: docs/Custom_Operation_for_GPUs.md [hunk 6]
```diff
@@ -558,543 +558,313 @@ The values have been computed correctly for forward operation, however, the gene
...
-emb_dim=512
-x = jax.random.normal(
-    jax.random.key(0),
-    shape=(jax.local_device_count() * per_core_batch_size, seq_len, emb_dim),
-    dtype=jnp.bfloat16,
-)
-norm_shape = x.shape[-2:]
-weight = jnp.ones(norm_shape, dtype=jnp.bfloat16)
-
+With that we have completely define our custom RMS norm primitive with custom_partitioning. To check for correctness we define the following loss functions: ref_loss is the reference value to compare against, while custom_p_loss uses our new primitive that implements custom_partitioning.
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:778]: define -> defined?

File: docs/Custom_Operation_for_GPUs.md [hunk 7]
```diff
@@ -558,543 +558,313 @@ The values have been computed correctly for forward operation, however, the gene
...
+Now there are no all-gathers in the HLO, sharding is respected and only gradients are accumulated via an all-reduce.
+
+
+## Let's put it together
+
+The complete definition of the primitives using custom_partitioning can be found in `Custom_Operation_for_GPUs.py`. All the cpp code used in this example is included below.
+
 ## Appendix
 
 ### `gpu_ops` code listing
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:870]: remove the old code here and add links to the files in the repo?

File: docs/Custom_Operation_for_GPUs.md (hunk 1 above)
Comment (keshavb96) [reply docs/Custom_Operation_for_GPUs.md:37]: It is mentioned a little later in the doc that xmap is deprecated. I will just remove it from here.

**Review by pschuh — APPROVED** [review]

**Review by superbobry — COMMENTED** [review]
General comment: should we perhaps choose a simpler op/ops for this? RMSNorm needs a lot of CUDA code which is not directly relevant here.

File: docs/Custom_Operation_for_GPUs.py [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+
+
+class RmsNormFwdClass:
+    name = "rms_forward_affine_mixed_dtype"
+    multiple_results = True
+    impl_static_args = (2,)    # eps
+    inner_primitive = None
+    outer_primitive = None
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (superbobry) [reply docs/Custom_Operation_for_GPUs.py:289]: Here is a sketch I did a few months ago

```
class UserPrimitive(abc.ABC):

  def __init_subclass__(cls, *, multiple_results: bool = False):
    if hasattr(cls, "_prim"):
      # This is a workaround for @dataclass re-creating the class.
      return

    prim = cls._prim = core.Primitive(cls.__name__.lower())
    prim.multiple_results = multiple_results
    prim.def_abstract_eval(
        lambda *avals, **params: cls(**params).abstract_eval(*avals)
    )
    prim.def_impl(partial(xla.apply_primitive, prim))

  def __call__(self, *args: Any) -> Any:
    return self.unwrap().bind(*args, **dataclasses.asdict(self))

  @abc.abstractclassmethod
  def abstract_eval(self, *args: core.ShapedArray) -> Sequence[core.ShapedArray]:
    ...

  # ...

@dataclasses.dataclass(frozen=True, slots=True)
class TopK(UserPrimitive, multiple_results=True):
  k: int
  # ...
```

**Comment by nouiz** [comment]
> General comment: should we perhaps choose a simpler op/ops for this? RMSNorm needs a lot of CUDA code which is not directly relevant here.

People don't need to read the CUDA code here.
If we took something simpler, we won't be able to show all features, like multi-output.

If we find something simpler that still show all features, that would be great. But I don't have an idea.
But this should be a follow up PR I think.

File: docs/gpu_ops/gpu_ops.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  dict["rms_forward_affine_mixed_dtype"] =
+      gpu_ops::EncapsulateFunction(gpu_ops::rms_forward_affine_mixed_dtypes);
+  dict["rms_backward_affine"] =
+      gpu_ops::EncapsulateFunction(gpu_ops::rms_backward_affine);
+  return dict;
+}
+
+PYBIND11_MODULE(gpu_ops, m) {
+  m.def("get_rms_norm_registrations", &RMSNormRegistrations);
+  m.def("create_rms_norm_descriptor",
```
[the remark this answers is not in the document]
Comment (nouiz) [reply docs/gpu_ops/gpu_ops.cpp:16]: I was told this isn't 100% finished. But when done, that would be great.

File: docs/Custom_Operation_for_GPUs.py [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+    """
+    register jax primitive
+
+    The order of calls. Each operation is composed of two primitives: Inner and Outer.
+
+    Inner, only the basic to wrap the custom_call itself.
+    - impl to XLA custom_call in C.
+    - abstract to know the static shapes
+    - lower to StableHLO XLA custom_call.
+    Outer, mostly all the rest:
```
[the remark this answers is not in the document]
Comment (nouiz) [reply docs/Custom_Operation_for_GPUs.py:248]: This is sadly a current hard requirement from JAX/XLA.
If we don't do that, there is crashes right now.
@mattjj can give more details here.
Hopefully we will remove that need later.
But right now, this is mostly hidden from the user in one function. So they don't need to understand all that.

**Comment by mjsML** [comment]
> > General comment: should we perhaps choose a simpler op/ops for this? RMSNorm needs a lot of CUDA code which is not directly relevant here.
> 
> People don't need to read the CUDA code here. If we took something simpler, we won't be able to show all features, like multi-output.
> 
> If we find something simpler that still show all features, that would be great. But I don't have an idea. But this should be a follow up PR I think.

@jaro-sevcik had similar perception I believe ... any ideas here?

File: docs/gpu_ops/gpu_ops.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (superbobry) [reply docs/gpu_ops/gpu_ops.cpp:16]: It is indeed not 100% finished e.g. not all data types have "casters" (in the pybind11 sense), but it is straightforward to add new ones.

CC @ezhulenev

File: docs/Custom_Operation_for_GPUs.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (superbobry) [reply docs/Custom_Operation_for_GPUs.py:248]: Thanks, Frédéric! I'm surprised to hear there are crashes if we don't do this exact implementation.

Should we discuss this in a bit more detail in Chat?

**Comment by superbobry** [comment]
> If we took something simpler, we won't be able to show all features, like multi-output.

What are some of the other features we want to demonstrate?

> If we find something simpler that still show all features, that would be great. But I don't have an idea. But this should be a follow up PR I think.

Yeah, I agree this could be done in a follow up.

File: docs/Custom_Operation_for_GPUs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (keshavb96) [reply docs/Custom_Operation_for_GPUs.py:289]: Can we get this merged without this change and then think about how to refactor / change the API?

[the remark this answers is not in the document]
Comment (superbobry) [reply docs/Custom_Operation_for_GPUs.py:289]: Okay, let's merge this as is.

**Comment by keshavb96** [comment]
Yes, I think it should be good to go

**Review by superbobry — APPROVED** [review]
I wonder if this example would be better of by being in a separate repo with tests making sure it works. 

The current version (and in fact the previous one as well) uses lots of JAX internal APIs which come with no stability guarantee, so it is pretty much guaranteed to break as time goes (unless we test it).

File: docs/Custom_Operation_for_GPUs.py [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+####################
+
+
+# Register functions defined in gpu_ops as custom call target for GPUs
+for _name, _value in gpu_ops.get_rms_norm_registrations().items():
+    xla_client.register_custom_call_target(_name, _value, platform="gpu")
+
+
+def element_type_to_descriptor_type_mapping(element_type):
+    _element_type_to_descriptor_type_mapping = {
```
[the remark this answers is not in the document]
Comment (jpienaar) [reply docs/Custom_Operation_for_GPUs.py:60]: I'm surprised this function isn't in Jax proper already. Seems like a GPU agnostic mapping.

[the remark this answers is not in the document]
Comment (superbobry) [reply docs/Custom_Operation_for_GPUs.py:60]: It's mapping to custom-op-specific enum members here.

I think another option is to pass the MLIR type as string and parse it on the C++ side?

**Comment by superbobry** [comment]
Can you squash the commits, please? I suspect that could be the reason for the import/copybara check failing.

File: docs/Custom_Operation_for_GPUs.md [hunk 12]
```diff
@@ -558,1657 +558,315 @@ The values have been computed correctly for forward operation, however, the gene
...
-            PartitionSpec("x", None, None),
-            PartitionSpec(None, None),
-        ),
-    )
-    out = pjitted(x, weight)
-
-for r, o in zip(ref, out):
-    print(jnp.allclose(r, o, atol=1e-2, rtol=1e-2))
+    def run_and_verify(loss):
+        pjitted = pjit(
```
Comment (nouiz) [inline docs/Custom_Operation_for_GPUs.md:796]: Would be great if this is converted to jit() as pjit() is deprecated. Can be a follow up PR too.

Comment (keshavb96) [reply docs/Custom_Operation_for_GPUs.md:796]: I will do this in a follow up.

**Comment by keshavb96** [comment]
@superbobry It still seems like copybara is failing despite squashing all the commits

**Comment by superbobry** [comment]
It looks like Copybara is unhappy that some files are missing the Apache license header. Can you make sure all files have one, please?

**Comment by superbobry** [comment]
Can you squash the commits, please?

**Comment by keshavb96** [comment]
Done!
